edgedb-client/src/builder.rs

```diff
@@ -1110,27 +1212,60 @@ impl Builder {
                 Run `edgedb project init` or use environment variables \
                 to configure connection."));
         }
+        let host = self.get_host();
+        if host.is_empty() {
+            return Err(ClientError::with_message("invalid host: empty string"));
+        } else if host.contains(",") {
+            return Err(ClientError::with_message("invalid host: multiple hosts"));
+        }
+        if self.get_user().is_empty() {
+            return Err(ClientError::with_message("invalid user: empty string"));
+        }
+        if self.get_database().is_empty() {
+            return Err(ClientError::with_message("invalid database: empty string"));
+        }
+        if self.get_port() == 0 {
+            return Err(ClientError::with_message("invalid port: 0"));
+        }
```

Comment:
I think it's better to move these to appropriate setters? We don't have error result in `set_host`/`set_port`/`set_database`, but should probably have. DSN can already fail so these checks there will be just fine.

Also these are probably `ClientInterfaceError`. I.e. invalid usages by user, right?

Reply:
> I think it's better to move these to appropriate setters?

I'm not 100% sure if this is right for a "builder", but it's done now.

> Also these are probably `ClientInterfaceError`. I.e. invalid usages by user, right?

Sure, `InvalidArgumentError` I guess.